Fix 'viirs_sdr' repeating data when TC geolocation was not available

djhoese opened the pull request:
The 'viirs_sdr' reader is meant to choose terrain corrected geolocation data if available or fall back to non-TC. You can control this with the `use_tc` kwarg where True means only use TC, False means only use non-TC, and None means use TC if available, non-TC otherwise. As described in #1786, if you have only non-TC files the reader will produce a DataArray with the correct number of rows, but the SwathDefinition in `.attrs['area']` will contain 3x the number of rows as it should.

This issue is caused by the geolocation determination step loading geolocation files regardless of whether or not they have been loaded before. I think this was not caught or has started being an issue for two reasons:

1. The tests weren't properly testing things (see below)
2. I think the `yaml_reader.py` file handler/filename stuff was optimized by me and the assumption was made that the file handlers are only ever created once. The viirs_sdr did not fit this assumption.

The tests were bad for two reasons:

1. They didn't check the same of the area against the shape of the data. Not a huge thing, but something that is important to check in hindsight.
2. They didn't create the on-disk geolocation files which are needed because the reader uses `glob` to find the geolocation filenames matching the glob pattern created from the data file's geo reference attribute.

 - [x] Closes #1786 <!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [x] Tests added <!-- for all bug fixes or enhancements -->

Possible alternative fixes:

1. Update the base reader to not create file handlers if they already exist. Given the assumption mentioned above this would really only benefit the VIIRS SDR reader (and maybe a couple others) for these special case situations.
2. Do fancier checking for "I want to load geo file X which is of file type FT, do I have that already? Yes? Then don't try to create any new file handlers".

mraspaud commented on satpy/tests/reader_tests/test_viirs_sdr.py:
```diff
@@ -480,19 +488,25 @@ def test_load_all_m_reflectances_provided_geo(self):
-                     'M10',
-                     'M11',
-                     ])
+        geo_fn1 = self._touch_geo_file("GMTCO")
```
Should we have a small helper context like:
```python
from contextlib import contextmanager
@contextmanager
def touch_geo_files(*prefixes):
    geofiles = [_touch_geo_file(prefix) for prefix in prefixes]
    try:
        yield
    finally:
        for filename in geofiles:
            os.remove(filename)
```

djhoese replied:
Done.

mraspaud reviewed (commented):
It's been a long time, so I don't really have in mind all the use cases, but this looks good to me.
The code should be refactored to make things clearer, but it could be kept for a later pr.
I just have a styling suggestion inline.

djhoese commented:
@mraspaud I took a stab at refactoring. The second half of the big for loop got much better and ended up realizing that the `sum` of the file handlers lists is no longer needed (the file handlers instance property are updated when the new geo files are loaded). However, the first half didn't seem like it was at the same abstraction l tried to refactor it into a separate method and it just feels gross. I think it feels that way because I don't know what `req` (required?) and `rem` are short for. Any idea?

mraspaud commented:
Required and remaining?

djhoese commented:
Ok so I renamed req/rem to primary/secondary since they are only ever a single string (not a "remaining" container of anything). I also simplified some of the related code. One specifically is that the handling of no geo filenames being found is now just handled by the code that would load new files. Since geo filenames will be an empty list then no files are loaded. I think this simplifies the code quite a bit. I'll merge this when I start work on Monday if you haven't already or don't have any other feedback.
